### PR TITLE
Calculate last uses while building intervals.

### LINKED ISSUE #7256: JIT throughput: Eliminate setLastUses
`LinearScan::setLastUses()` was originally added because the ordering of local variable references within a statement was not always correct. I believe that those bugs have been fixed, so that this should be changed to a debug-only phase that validates that the last use info is correct.

### PR DESCRIPTION
**Pull request by pgavlin**
This change moves LSRA's last use calculation from `setLastUses`
into `buildIntervals`.

The algorithm flips is as follows:
- When adding a ref position to a lclVar interval:
    - If the ref position is not a parameter def, zero init, or an
      exposed use, set the ref position's last use bit
    - Additionally, if the ref position is a use and there is a
      preceding ref position in the same basic block, clear that ref
      position's last use bit.
- Once a basic block has been processed, clear the last use bit of the
  final ref position in that basic block (if any exists) for each lclVar
  that is live out of that block.

Aside from the algorithmic change, the other major difference is that
the GTF_VAR_DEATH bit is not updated until `resolveLocalRef`.

Fixes #7256.

### CODE REVIEW

**Comment by pgavlin** [comment]
@sivarv @BruceForstall @russellhadley @dotnet/jit-contrib PTAL.

There are no assembly diffs due to this change, and local testing shows a modest throughput improvement.

File: src/jit/lsra.cpp [hunk 1]
```diff
@@ -3883,31 +3875,28 @@ void LinearScan::buildRefPositionsForNode(GenTree*                  tree,
...
-        if (isCandidateLocalRef(useNode))
+        regMaskTP fixedAssignment = fixedCandidateMask(type, candidates);
+        if (fixedAssignment != RBM_NONE)
         {
-            isLastUse = ((useNode->gtFlags & GTF_VAR_DEATH) != 0);
+            candidates = fixedAssignment;
         }
-        else
+
+        const bool regOptionalAtUse = useNode->IsRegOptional();
```
Comment (sandreenko) [inline src/jit/lsra.cpp]: Is regOptopnalAtUse used once? Is it possible to move it  to 3960?

Comment (pgavlin) [inline src/jit/lsra.cpp]: Not without some other refactoring: its value is derived from `useNode`, which we set to `nullptr` on [line 3897](https://github.com/dotnet/coreclr/pull/10032/files#diff-93fb9832a03a9783e2a415b581fcb45cR3897) before the use of `regOptionalAtUse` on [line 3960](https://github.com/dotnet/coreclr/pull/10032/files#diff-93fb9832a03a9783e2a415b581fcb45cR3960).

**Comment by pgavlin** [comment]
@adiaaida: how do I run a private throughput job with this change?

**Comment by adiaaida** [comment]
@pgavlin Yes. It's test Windows_NT <arch> throughput. Arch options are x64, x86, x86jit32.

**Comment by pgavlin** [comment]
@dotnet-bot test Windows_NT x64 throughput
@dotnet-bot test Windows_NT x86 throughput

**Comment by adiaaida** [comment]
Sorry, underscores.

@dotnet-bot test Windows_NT_x64 throughput
@dotnet-bot test Windows_NT_x86 throughput

**Comment by pgavlin** [comment]
Is that easy to change? If so, can we remove the underscores for consistency with the rest of the request strings?

**Comment by adiaaida** [comment]
Yes, that's an easy change. I copied what we did for the perf jobs (they also have an underscore between os and arch).

**Comment by pgavlin** [comment]
@dotnet-bot test OSX x64 Checked Build and Test

**Comment by pgavlin** [comment]
Ping. Is anybody available to take a look at this?

**Comment by sivarv** [comment]
I am taking a look at it right now

**Comment by sivarv** [comment]
Please trigger full set of LSRA stress runs on x64 (windows and ubuntu) and x86 windows to make sure there are no regressions.

**Comment by pgavlin** [comment]
@dotnet-bot test Windows_NT jitstressregs1
@dotnet-bot test Windows_NT jitstressregs2
@dotnet-bot test Windows_NT jitstressregs3
@dotnet-bot test Windows_NT jitstressregs4
@dotnet-bot test Windows_NT jitstressregs8
@dotnet-bot test Windows_NT jitstressregs0x1000
@dotnet-bot test Windows_NT jitstressregs0x80
@dotnet-bot test Windows_NT jitstressregs0x10

File: src/jit/lsra.cpp [hunk 2]
```diff
@@ -2486,16 +2498,15 @@ RefType refTypeForLocalRefNode(GenTree* node)
 // being set by dataflow analysis.  It is necessary to do it this way only because the execution
 // order wasn't strictly correct.
```
Comment (sivarv) [inline src/jit/lsra.cpp]: This comment block needs to be updated, given its modified functionality

**Comment by sivarv** [comment]
Since last use info impacts gcinfo, it would be a good idea to trigger a gc-stress run as well

**Comment by sivarv** [comment]
Not whether we have gcinfo diff similar to asm diff to verify that gcinfo reporting didn't change in unexpected ways.

**Comment by sivarv** [comment]
:shipit:

**Comment by pgavlin** [comment]
> Not whether we have gcinfo diff similar to asm diff to verify that gcinfo reporting didn't change in unexpected ways.

Sure. FWIW, the statement about "no asm diffs" above does apply to GC info as well: SPMI reported no diffs in any of the bits output by the JIT (code, GC info, EH info, etc.).

**Comment by pgavlin** [comment]
@dotnet-bot test Windows_NT gcstress0xc

**Comment by pgavlin** [comment]
The GC stress failures are packaging-related, but it does look like there are some failures in the JitStressRegs jobs that I'll need to investigate.

**Comment by pgavlin** [comment]
@dotnet-bot test Windows_NT jitstressregs1
@dotnet-bot test Windows_NT jitstressregs2
@dotnet-bot test Windows_NT jitstressregs3
@dotnet-bot test Windows_NT jitstressregs4
@dotnet-bot test Windows_NT jitstressregs8
@dotnet-bot test Windows_NT jitstressregs0x1000
@dotnet-bot test Windows_NT jitstressregs0x80
@dotnet-bot test Windows_NT jitstressregs0x10

**Comment by pgavlin** [comment]
@dotnet-bot test Windows_NT gcstress0xc

**Comment by pgavlin** [comment]
@dotnet-bot test Ubuntu jitstressregs1
@dotnet-bot test Ubuntu jitstressregs2
@dotnet-bot test Ubuntu jitstressregs3
@dotnet-bot test Ubuntu jitstressregs4
@dotnet-bot test Ubuntu jitstressregs8
@dotnet-bot test Ubuntu jitstressregs0x1000
@dotnet-bot test Ubuntu jitstressregs0x80
@dotnet-bot test Ubuntu jitstressregs0x10

@dotnet-bot test OSX jitstressregs1
@dotnet-bot test OSX jitstressregs2
@dotnet-bot test OSX jitstressregs3
@dotnet-bot test OSX jitstressregs4
@dotnet-bot test OSX jitstressregs8
@dotnet-bot test OSX jitstressregs0x1000
@dotnet-bot test OSX jitstressregs0x80
@dotnet-bot test OSX jitstressregs0x10

**Comment by pgavlin** [comment]
The GC stress failures are timeouts that we see occasionally.

**Comment by pgavlin** [comment]
@dotnet-bot test OSX jitstressregs4
@dotnet-bot test OSX jitstressregs8

**Comment by pgavlin** [comment]
@dotnet-bot test Windows_NT gcstress0xc

**Pull request merged**